[SPARK-31069][CORE] Avoid repeat compute `chunksBeingTransferred` cause hight cpu cost in external shuffle service when  `maxChunksBeingTransferred`  use default value.

AngersZhuuuu opened the pull request:
### What changes were proposed in this pull request?
Followup from #27831 , origin author @chrysan.

Each request it will check `chunksBeingTransferred `
```
public long chunksBeingTransferred() {
    long sum = 0L;
    for (StreamState streamState: streams.values()) {
      sum += streamState.chunksBeingTransferred.get();
    }
    return sum;
  }
```
  such as 
```
long chunksBeingTransferred = streamManager.chunksBeingTransferred();
    if (chunksBeingTransferred >= maxChunksBeingTransferred) {
      logger.warn("The number of chunks being transferred {} is above {}, close the connection.",
        chunksBeingTransferred, maxChunksBeingTransferred);
      channel.close();
      return;
    }
```
It will  traverse `streams` repeatedly and we know that fetch data chunk will access `stream` too,  there cause two problem:

1. repeated traverse `streams`, the longer the length, the longer the time
2. lock race in ConcurrentHashMap `streams`

In this PR, when `maxChunksBeingTransferred` use default value, we avoid compute `chunksBeingTransferred ` since we don't  care about this.  If user want to set this configuration and meet performance problem,  you can also backport PR #27831 


### Why are the changes needed?
Speed up  getting `chunksBeingTransferred`  and avoid lock race in object `streams`


### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Existed UT

AngersZhuuuu committed:
commit

SparkQA commented:
**[Test build #130198 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130198/testReport)** for PR 30139 at commit [`6857e4e`](https://github.com/apache/spark/commit/6857e4e7eed637da47e2878a3af40ae62ff60bd8).

AngersZhuuuu commented:
ping @wangyum @dongjoon-hyun @HeartSaVioR @jiangxb1987 @Ngone51 
With path  cost duration grows linearly in `streams`'s length.  The results seem to be in line with our expectations.

wangyum commented:
Thank you @AngersZhuuuu' number. We have been running this change for more than half a year.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34799/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34799/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34799/
Test PASSed.

AngersZhuuuu committed:
Update OneForOneStreamManager.java

SparkQA commented:
**[Test build #130199 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130199/testReport)** for PR 30139 at commit [`72e7b87`](https://github.com/apache/spark/commit/72e7b87e7b2a4a6b0925b1daf6bc600efabebd4b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34800/

SparkQA commented:
**[Test build #130198 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130198/testReport)** for PR 30139 at commit [`6857e4e`](https://github.com/apache/spark/commit/6857e4e7eed637da47e2878a3af40ae62ff60bd8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130198/
Test PASSed.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34800/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34800/
Test PASSed.

SparkQA commented:
**[Test build #130199 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130199/testReport)** for PR 30139 at commit [`72e7b87`](https://github.com/apache/spark/commit/72e7b87e7b2a4a6b0925b1daf6bc600efabebd4b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130199/
Test PASSed.

mridulm commented:
+CC @otterc

dongjoon-hyun commented:
Could you add an empty commit whose authorship is the original author, @AngersZhuuuu . If then, Apache Spark merge script can give both of you the authorship properly.

chrysan committed:
Set original author

SparkQA commented:
**[Test build #130241 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130241/testReport)** for PR 30139 at commit [`0654684`](https://github.com/apache/spark/commit/0654684e47a826d028ed4dc1e2199bda697d0f37).

AngersZhuuuu commented:
> Could you add an empty commit whose authorship is the original author, @AngersZhuuuu . If then, Apache Spark merge script can give both of you the authorship properly.

Done this. thanks for mention this point.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34841/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34841/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34841/
Test PASSed.

SparkQA commented:
**[Test build #130241 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130241/testReport)** for PR 30139 at commit [`0654684`](https://github.com/apache/spark/commit/0654684e47a826d028ed4dc1e2199bda697d0f37).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130241/
Test PASSed.

mridulm reviewed (commented):
I am concerned that the `totalChunksBeingTransferred` can diverge over time from state of `streams` when there are concurrent updates. Either updates to both should be within the same critical section, or we should carefully ensure there is no potential for divergence. Can you relook and ensure this is not the case ?

For example, can there be concurrent execution between `chunkBeingSent`, `chunkSent` and `connectionTerminated` ? If yes, can we ensure the state remains consistent.

AngersZhuuuu commented:
> I am concerned that the `totalChunksBeingTransferred` can diverge over time from state of `streams` when there are concurrent updates. Either updates to both should be within the same critical section, or we should carefully ensure there is no potential for divergence. Can you relook and ensure this is not the case ?
> 
> For example, can there be concurrent execution between `chunkBeingSent`, `chunkSent` and `connectionTerminated` ? If yes, can we ensure the state remains consistent. 

In origin code, we just use a concurrent hash map(streams) and also there no strong locking mechanism. And `chunksBeingTransferred` is only used to check with config.
```
    long chunksBeingTransferred = streamManager.chunksBeingTransferred();
    if (chunksBeingTransferred >= maxChunksBeingTransferred) {
      logger.warn("The number of chunks being transferred {} is above {}, close the connection.",
        chunksBeingTransferred, maxChunksBeingTransferred);
      channel.close();
      return;
    }
```

IMO, we can add a lock to keep strong consistence of value `totalChunksBeingTransferred`, such as 
![image](https://user-images.githubusercontent.com/46485123/97126839-f2abfb80-1772-11eb-9271-54775defde35.png)

And I have run the test in desc,  this change has little effect on performance. WDYT?

jiangxb1987 commented:
Yes we should ensure the streamState and the totalChunksBeingTransfered are updated synchronically. Other than that the PR looks good!

jiangxb1987 commented on common/network-common/src/main/java/org/apache/spark/network/server/OneForOneStreamManager.java:
```diff
@@ -43,6 +43,7 @@
 
   private final AtomicLong nextStreamId;
   private final ConcurrentHashMap<Long, StreamState> streams;
+  private final AtomicLong totalChunksBeingTransferred = new AtomicLong(0);
```
nit: maybe rename to `numChunksBeingTransferred`? Because it's not accumulating all the chunks that are transferred in history.

AngersZhuuuu replied:
> nit: maybe rename to `numChunksBeingTransferred`? Because it's not accumulating all the chunks that are transferred in history.

Updated

otterc commented:
> IMO, we can add a lock to keep strong consistence of value `totalChunksBeingTransferred`, such as
> ![image](https://user-images.githubusercontent.com/46485123/97126839-f2abfb80-1772-11eb-9271-54775defde35.png)
> 
> And I have run the test in desc, this change has little effect on performance. WDYT?

This should have a considerable impact on the performance when there are multiple open streams because updates of different streams would lock on `totalChunksBeingTransferred`?

AngersZhuuuu commented:
> This should have a considerable impact on the performance when there are multiple open streams because updates of different streams would lock on a single object `totalChunksBeingTransferred`. Isn't that the case?

Since `totalChunksBeingTransfereed` is atomic, when it update It has its own competition, add a lock at `totalChunksBeingTransfereed` won't have too much impact. And we can keep strong consistences between the streamState and the totalChunksBeingTransfered are updated synchronically.

otterc commented:
> > This should have a considerable impact on the performance when there are multiple open streams because updates of different streams would lock on a single object `totalChunksBeingTransferred`. Isn't that the case?
> 
> Since `totalChunksBeingTransfereed` is atomic, when it update It has its own competition, add a lock at `totalChunksBeingTransfereed` won't have too much impact. And we can keep strong consistences between the streamState and the totalChunksBeingTransfered are updated synchronically.

Hmmm. Every update to `chunkSent` and `chunkBeingSent` will compete for the lock on object `totalChunksBeingTransferred` if we add the `synchronize(totalChunksBeingTransferred)`. This would increase the time for these operations. This would mean that to speed up `chunksBeingTransferred`, we are increasing the time of updates to `streamState`.

AngersZhuuuu committed:
Update OneForOneStreamManager.java

AngersZhuuuu commented:
> Yes we should ensure the streamState and the totalChunksBeingTransfered are updated synchronically. Other than that the PR looks good!

How about current change? Don't use AtomicLong but use `synchronize` to  keep  strong consistency. And the test result is 
```
OneForOneStreamManager fetch data duration test:
Stream Size          Max           Min           Avg
10000           1796           187           497.5
50000           4214           1295           2267.9
100000           10635           3643           5800.3

Process finished with exit code 0

```

SparkQA commented:
**[Test build #130316 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130316/testReport)** for PR 30139 at commit [`a32d6f3`](https://github.com/apache/spark/commit/a32d6f31e285c767825545aa5c3033b08bd2f271).

AngersZhuuuu commented:
> Hmmm. Every update to `chunkSent` and `chunkBeingSent` will compete for the lock on object `totalChunksBeingTransferred` if we add the `synchronize(totalChunksBeingTransferred)`. This would increase the time for these operations. This would mean that to speed up `chunksBeingTransferred`, we are increasing the time of updates to `streamState`.

I know that, but as @jiangxb1987  @mridulm mentioned, we need to  ensure the streamState and the totalChunksBeingTransfered are updated synchronically.  Add this lock is a strong guarantee.  The execution process in the middle of the lock is very fast, so the impact is not really significant

AngersZhuuuu commented:
> Hmmm. Every update to `chunkSent` and `chunkBeingSent` will compete for the lock on object `totalChunksBeingTransferred`

We reduce many race condition on `streams` and just add a very quick lock  on. `numChunksBeingTransferred`

HeartSaVioR commented:
Sorry but your latest change doesn't actually lock properly. Long is immutable, and you always replace the object when you do the calculation, which is used as a lock. Your best try would be changing it to long, and have a separate Object field as locking purpose.

AngersZhuuuu commented:
> Sorry but your latest change doesn't actually lock properly. Long is immutable, and you always replace the object when you do the calculation and assign to the field, which is used as a lock. Your best try would be changing it to long, and have a separate Object field as locking purpose.

a  big mistake, thanks for your suggestion.

AngersZhuuuu committed:
Update OneForOneStreamManager.java

SparkQA commented:
**[Test build #130319 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130319/testReport)** for PR 30139 at commit [`b23dbd0`](https://github.com/apache/spark/commit/b23dbd03d0f4219ab8db3fa3557d7a50634821f1).

SparkQA commented:
**[Test build #130316 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130316/testReport)** for PR 30139 at commit [`a32d6f3`](https://github.com/apache/spark/commit/a32d6f31e285c767825545aa5c3033b08bd2f271).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130316/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34918/

AngersZhuuuu commented:
> Sorry but your latest change doesn't actually lock properly. Long is immutable, and you always replace the object when you do the calculation and assign to the field, which is used as a lock. Your best try would be changing it to long (to avoid box/unbox), and have a separate Object field as locking purpose.

Update and check 

**_1. with UT change_**
![image](https://user-images.githubusercontent.com/46485123/97267072-84dffc80-1864-11eb-8125-9dc05b92886b.png)

Current change With lock 
```
OneForOneStreamManager fetch data duration test:
Stream Size          Max           Min           Avg
10000           2834           704           1188.0
50000           14048           10789           12055.7
```

Only use AtomicLong
```
OneForOneStreamManager fetch data duration test:
Stream Size          Max           Min           Avg
10000           6723           724           1527.5
50000           14673           10712           12080.5
```

**_2. Without UT change_**
Current change With lock 
```
OneForOneStreamManager fetch data duration test:
Stream Size          Max           Min           Avg
10000           822           176           360.1
50000           4190           1263           2387.9
100000           6785           3870           4811.9
```
Only use AtomicLong
```
OneForOneStreamManager fetch data duration test:
Stream Size          Max           Min           Avg
10000           1037           167           548.5
50000           2982           1102           1712.8
100000           8779           3165           4831.9
```

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34918/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34918/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34921/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34921/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34921/
Test PASSed.

SparkQA commented:
**[Test build #130319 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130319/testReport)** for PR 30139 at commit [`b23dbd0`](https://github.com/apache/spark/commit/b23dbd03d0f4219ab8db3fa3557d7a50634821f1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130319/
Test FAILed.

AngersZhuuuu commented:
retest this please

SparkQA commented:
**[Test build #130327 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130327/testReport)** for PR 30139 at commit [`b23dbd0`](https://github.com/apache/spark/commit/b23dbd03d0f4219ab8db3fa3557d7a50634821f1).

SparkQA commented:
**[Test build #130327 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130327/testReport)** for PR 30139 at commit [`b23dbd0`](https://github.com/apache/spark/commit/b23dbd03d0f4219ab8db3fa3557d7a50634821f1).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130327/
Test FAILed.

AngersZhuuuu committed:
Update OneForOneStreamManager.java

SparkQA commented:
**[Test build #130328 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130328/testReport)** for PR 30139 at commit [`a4b70bf`](https://github.com/apache/spark/commit/a4b70bf65ea08e91b32376f415abac7110010abb).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34928/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34930/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34928/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34928/
Test PASSed.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34930/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34930/
Test PASSed.

SparkQA commented:
**[Test build #130328 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130328/testReport)** for PR 30139 at commit [`a4b70bf`](https://github.com/apache/spark/commit/a4b70bf65ea08e91b32376f415abac7110010abb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130328/
Test PASSed.

srowen commented on common/network-common/src/main/java/org/apache/spark/network/server/OneForOneStreamManager.java:
```diff
@@ -43,6 +43,8 @@
 
   private final AtomicLong nextStreamId;
   private final ConcurrentHashMap<Long, StreamState> streams;
+  private Long numChunksBeingTransferred = 0L;
```
Why not `long`?

AngersZhuuuu replied:
> Why not `long`?

emmm, changed

srowen commented on common/network-common/src/main/java/org/apache/spark/network/server/OneForOneStreamManager.java:
```diff
@@ -123,7 +125,10 @@ public void connectionTerminated(Channel channel) {
       if (state.associatedChannel == channel) {
-        streams.remove(entry.getKey());
+        synchronized (lock) {
+          streams.remove(entry.getKey());
```
Is locking on `streams` not sufficient?

AngersZhuuuu replied:
> Is locking on `streams` not sufficient?
 
Haven't thought of a particularly good solution to consistency and performance conflicts

srowen replied:
I just mean you could use it as the mutex but maybe that has undesirable interactions with how it synchronizes other access. OK no big deal.

AngersZhuuuu replied:
> I just mean you could use it as the mutex but maybe that has undesirable interactions with how it synchronizes other access. OK no big deal.

Yea

AngersZhuuuu committed:
Update OneForOneStreamManager.java

SparkQA commented:
**[Test build #130334 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130334/testReport)** for PR 30139 at commit [`53b3c0b`](https://github.com/apache/spark/commit/53b3c0bb90d4b22e4994ea44a659026c4ba0c93c).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34936/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34936/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34936/
Test FAILed.

otterc commented:
>  I know that, but as @jiangxb1987 @mridulm mentioned, we need to ensure the streamState and the totalChunksBeingTransfered are updated synchronically. Add this lock is a strong guarantee. The execution process in the middle of the lock is very fast, so the impact is not really significant. Also, It's a huge leap in performance compared to what it was before

Speeding up `chunksBeingTransferred` while ensuring strong consistency with introducing this lock on an update of every streamstate has a cost. The cost here is that we are increasing the time it to takes to update the streamstate. In a production like environment with a lot of concurrent streams, this will increase the time of `chunkSent` and `chunkBeingSent`

> The execution process in the middle of the lock is very fast, so the impact is not really significant. 

With this strong consistency enforced, have you scale tested it? I haven't run this change, but I am quite sure that the impact on `chunkSent` and `chunkBeingSent` will be significant.

> Also, It's a huge leap in performance compared to what it was before

The improvement in performance is on read of `chunksBeingTransferred`. But, as above earlier when you add strong consistency by introducing this lock it will degrade the performance of `chunkSent` and `chunkBeingSent`.

SparkQA commented:
**[Test build #130334 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130334/testReport)** for PR 30139 at commit [`53b3c0b`](https://github.com/apache/spark/commit/53b3c0bb90d4b22e4994ea44a659026c4ba0c93c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130334/
Test FAILed.

AngersZhuuuu commented:
retest this please

SparkQA commented:
**[Test build #130342 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130342/testReport)** for PR 30139 at commit [`53b3c0b`](https://github.com/apache/spark/commit/53b3c0bb90d4b22e4994ea44a659026c4ba0c93c).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34944/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34944/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34944/
Test FAILed.

SparkQA commented:
**[Test build #130342 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130342/testReport)** for PR 30139 at commit [`53b3c0b`](https://github.com/apache/spark/commit/53b3c0bb90d4b22e4994ea44a659026c4ba0c93c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130342/
Test PASSed.

AngersZhuuuu commented:
@otterc  Know what you are concern,  I haven't thought of such a good solution for a while
ping @jiangxb1987 @HeartSaVioR @dongjoon-hyun  Any more advise?

HeartSaVioR commented:
I somehow revisited the related configuration `spark.shuffle.maxChunksBeingTransferred` and realized the default value is set to `Long.MAX_VALUE`. (effectively off by default)

That leads me to wonder, does the value (`numChunksBeingTransferred`) need to be accurate at the specific time? Do we really need to make both updates be in same critical section? Can we tolerate the skew and allow eventual consistent of the value?

For the concern of being diverged, I guess updates to both would happen anyway unless there's no bug (I agree this is a tricky one to guarantee), as we anyway do the update on both sequentially. Please correct me if I'm missing something.

AngersZhuuuu commented:
> That leads me to wonder, does the value (`numChunksBeingTransferred`) need to be accurate at the specific time? Do we really need to make both updates be in same critical section? Can we tolerate the skew and allow eventual consistent of the value?
> (That said, the initial version of the patch looks OK for me.)

Agree with this, since this check just gives the customer a configurable number of the largest chunks that are being transferred. WDYT @jiangxb1987 @dongjoon-hyun .

mridulm commented:
If it will be eventually consistent, that is fine: what we can’t have is
permanent divergence (positive or to negative).
The current situation does exhibit eventual consistency (if updates while
computing, then it diverges from reality - but that is transient).

If the value will never converge, that is a problem.
The current pr, last I checked, has potential for this and that needs to be
fixed.



On Thu, Oct 29, 2020 at 2:14 AM Jungtaek Lim <notifications@github.com>
wrote:

> I somehow revisited the related configuration
> spark.shuffle.maxChunksBeingTransferred and realized the default value is
> set to Long.MAX_VALUE. (effectively off by default)
>
> That leads me to wonder, does the value (numChunksBeingTransferred) need
> to be accurate at the specific time? Do we really need to make both updates
> be in same critical section? Can we tolerate the skew and allow eventual
> consistent of the value?
>
> For the concern of being diverged, I guess updates to both would happen
> anyway unless there's no bug (I agree this is a tricky one to guarantee),
> as we anyway do the update on both sequentially. Please correct me if I'm
> missing something.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/30139#issuecomment-718410305>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMETFEYOUBJFR2AAVQL4BLSNEI43ANCNFSM4S4G3CXQ>
> .
>

HeartSaVioR commented:
Thanks, I took a look again, and `connectionTerminated` might be problematic indeed if other operations from the connection is taking place. (I don't know whether the case is realistic in practice, so that's a kind of hypothetically thinking.)

HeartSaVioR commented:
The root issue here is that `numChunksBeingTransferred` is called quite often.

If I understand correctly, the counter in stream is either increased or decreased via `processFetchRequest` & `processStreamRequest`, which `chunksBeingTransferred` is also called in prior. That said, the number of operations across counters would be just 2x of the number of calls for chunksBeingTransferred, and the difference of cost is significant. (increase/decrease atomic integer vs iterate all stream entities and read the atomic integer, and sum up) The cost for latter is linearly increasing based on the number of entities in streams, hence the problem appears if the number of entities is quite huge.

So that sounds like a trade-off. Probably it would be pretty much beneficial to leave it as it is if we assume the number of stream entities will retain small enough, but if we assume the case where the number of stream entities are quite large, the cost of synchronizing numChunksBeingTransferred is going to be smaller.

Possible alternative would be reducing the calculation of numChunksBeingTransferred - use cached value and update on condition like rate (once in 5 calls) or interval (update if the cached value is calculated earlier than XX seconds prior). We agreed that the value of numChunksBeingTransferred doesn't need to be strictly accurate, so this might be acceptable.

Would like to hear the voices. Thanks!

AngersZhuuuu commented:
> The cost for latter is linearly increasing based on the number of entities in streams, hence the problem appears if the number of entities is quite huge.

Each iterate of all streams is small, but each chunk will trigger count `chunksBeingTransferred `, the cost will be  very serious
and there is also a competition for `streams`'s locks since it is a ConcurrentHashMap.

> So that sounds like a trade-off. Probably it would be pretty much beneficial to leave it as it is if we assume the number of stream entities will retain small enough, but if we assume the case where the number of stream entities are quite large, the cost of synchronizing numChunksBeingTransferred is going to be smaller.

Yea, that is why I said  the cost of `synchronizing numChunksBeingTransferred` is  acceptable. 

> We agreed that the value of numChunksBeingTransferred doesn't need to be strictly accurate, so this might be acceptable.

From this point, I think the initial approach is good, the code is simple and easy to understand.

Ngone51 commented:
> Each iterate of all streams is small, but each chunk will trigger count chunksBeingTransferred , the cost will be very serious
and there is also a competition for streams's locks since it is a ConcurrentHashMap.

It doesn't make sense to me if we count the cost for the total cost of chunks. And your benchmark seems to prove that the main cost comes from the traversing of streams, isn't it?

BTW, shall we avoid traversing when `maxChunksBeingTransferred=Long.MAX_VALUE`?

AngersZhuuuu commented:
> It doesn't make sense to me if we count the cost for the total cost of chunks. And your benchmark seems to prove that the main cost comes from the traversing of streams, isn't it?

Yea.

> From this point of view, I think the initial approach is good, the code is simple and easy to understand

a good suggestion.   or most users, this parameter is not specified

Ngone51 commented:
Another thought is, would it ease the problem if we track the transferring chunks per channel, e.g., Map[Channle, long]?
For a single server, there's shouldn't be too many channels and the counter should be thread-safe per channel.

AngersZhuuuu commented:
> there's shouldn't be too many channels and the counter should be thread-safe per channel.

Need other professional member  to check whether the channel is in a controllable range.  IMO, When the data shuffle partition is large, I don't think the channel size will be small

AngersZhuuuu committed:
follow commebt

AngersZhuuuu committed:
Merge branch 'SPARK-31069' of https://github.com/AngersZhuuuu/spark into SPARK-31069

SparkQA commented:
**[Test build #130655 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130655/testReport)** for PR 30139 at commit [`d5fe461`](https://github.com/apache/spark/commit/d5fe46182bf39c127f971ab63ed7840fbedce227).

Ngone51 commented:
> IMO, When the data shuffle partition is large, I don't think the channel size will be small

https://github.com/apache/spark/blob/fa5dc0a45a414c34b31c5d7efe396aa04f1e66e3/common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java#L106-L109

The default concurrent connections between the peer are 1. So the total connection number of a single server should depend on the number of executors, right? If there are more than 5k or 10k executors, the problem does still exists.

@wangyum May I ask what's the scale of the cluster in your case?

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35266/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35266/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35266/
Test FAILed.

SparkQA commented:
**[Test build #130655 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130655/testReport)** for PR 30139 at commit [`d5fe461`](https://github.com/apache/spark/commit/d5fe46182bf39c127f971ab63ed7840fbedce227).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130655/
Test PASSed.

HeartSaVioR commented:
> Another thought is, would it ease the problem if we track the transferring chunks per channel, e.g., Map[Channle, long]?

I actually thought about this and looked into the logic, but we have a case where we remove the stream from streams "without disconnecting channel", which looks like we still need to know the count per stream.

https://github.com/apache/spark/blob/master/common/network-common/src/main/java/org/apache/spark/network/server/OneForOneStreamManager.java#L91

HeartSaVioR commented:
> BTW, shall we avoid traversing when maxChunksBeingTransferred=Long.MAX_VALUE?

This is a good idea to optimize for the default value. Thanks!

AngersZhuuuu commented:
> I actually thought about this and looked into the logic, but we have a case where we remove the stream from streams "without disconnecting channel", which looks like we still need to know the count per stream.

By the way,  this logic's performance is not good when channel size is large.

AngersZhuuuu commented:
> The default concurrent connections between the peer are 1. So the total connection number of a single server should depend on the number of executors, right? If there are more than 5k or 10k executors, the problem does still exists.

Yea, I've been in touch with users who set up thousands of Executors

AngersZhuuuu commented:
Any other Suggestions？ also ping @cloud-fan

cloud-fan commented on common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java:
```diff
@@ -88,12 +88,14 @@ public void processFetchRequest(
-        chunksBeingTransferred, maxChunksBeingTransferred);
-      channel.close();
-      return;
+    if (maxChunksBeingTransferred < Long.MAX_VALUE) {
```
Do we need this? `AtomicLong` seems efficient enough to always call `chunksBeingTransferred`

AngersZhuuuu replied:
> Do we need this? `AtomicLong` seems efficient enough to always call `chunksBeingTransferred`

OK,  how do you think about consistency and numerical convergence https://github.com/apache/spark/pull/30139#issuecomment-718528130

cloud-fan replied:
If the concern is we may miss some places that update the state, then maybe only keeping this change is safer.

AngersZhuuuu replied:
> If the concern is we may miss some places that update the state, then maybe only keeping this change is safer.

OK, so I will just revert change about  `if (maxChunksBeingTransferred < Long.MAX_VALUE) {`

cloud-fan replied:
Shouldn't we revert the change of `AtomicLong` and only keep `if (maxChunksBeingTransferred < Long.MAX_VALUE)`?

AngersZhuuuu replied:
> Shouldn't we revert the change of `AtomicLong` and only keep `if (maxChunksBeingTransferred < Long.MAX_VALUE)`?

All right,  I didn't get your point of comment https://github.com/apache/spark/pull/30139/files#r519647116, update  now

AngersZhuuuu replied:
ping @cloud-fan Any more suggestion?

AngersZhuuuu committed:
Update OneForOneStreamManager.java

SparkQA commented:
**[Test build #130779 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130779/testReport)** for PR 30139 at commit [`b07b999`](https://github.com/apache/spark/commit/b07b9997888d31e74ba86cc38c69c1d25ea23dca).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35388/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35388/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35388/
Test PASSed.

SparkQA commented:
**[Test build #130779 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130779/testReport)** for PR 30139 at commit [`b07b999`](https://github.com/apache/spark/commit/b07b9997888d31e74ba86cc38c69c1d25ea23dca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130779/
Test PASSed.

HeartSaVioR commented:
I see this is now only making exception on not calculating the chunksBeingTransferred on default value of config. This is far from the original approach, so could you please update PR title and description based on the new change?

IMHO we no longer fixes SPARK-31069 via this PR, but if there's no good idea to address the root issue, it'd be OK to keep this to be associated with SPARK-31069. If someone encounters with the issue setting the config as non-default, the issue would be filed again and we can work on the new JIRA issue.

AngersZhuuuu commented:
> I see this is now only making exception on not calculating the chunksBeingTransferred on default value of config. This is far from the original approach, so could you please update PR title and description based on the new change?
> 
> IMHO we no longer fixes [SPARK-31069](https://issues.apache.org/jira/browse/SPARK-31069) via this PR, but if there's no good idea to address the root issue, it'd be OK to keep this to be associated with [SPARK-31069](https://issues.apache.org/jira/browse/SPARK-31069). If someone encounters with the issue setting the config as non-default, the issue would be filed again and we can work on the new JIRA issue.

Yea, updated and  added guidelines for users who need to use the origin  PR #27831

mridulm commented:
@AngersZhuuuu To clarify, the updated behavior of the PR is that by default we dont incur the cost of computing `chunksBeingTransferred` as default values disable `maxChunksBeingTransferred`. But if `maxChunksBeingTransferred` is configured, the performance characteristics is same as earlier.

I am fine with the change, ideally we should fix the behavior even when `maxChunksBeingTransferred` is configured, but that can be a followup work.

AngersZhuuuu commented:
> I am fine with the change, ideally we should fix the behavior even when `maxChunksBeingTransferred` is configured, but that can be a followup work.

Yea, need to think about how to guarantee performance and consistent execution with minimal overhead as followup work. Current change will solve most users' problems.

mridulm commented:
+CC @srowen, @jiangxb1987, @cloud-fan 
I will leave it around till monday in case others want to comment, and then merge.

mridulm commented:
+CC @otterc, @wangyum, @HeartSaVioR, @Ngone51  who also reviewed this change.

HeartSaVioR reviewed (approved):
+1 Looks OK to me.

It'd be ideal if we could find the good way we all agree to resolve the origin problem, but it doesn't look like that easy.

otterc commented on common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java:
```diff
@@ -88,12 +88,14 @@ public void processFetchRequest(
+      long chunksBeingTransferred = streamManager.chunksBeingTransferred();
+      if (chunksBeingTransferred >= maxChunksBeingTransferred) {
+        logger.warn("The number of chunks being transferred {} is above {}, close the connection.",
+            chunksBeingTransferred, maxChunksBeingTransferred);
```
Nit: indentation should be 2

AngersZhuuuu replied:
> Nit: indentation should be 2

Done

otterc commented on common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java:
```diff
@@ -124,12 +124,14 @@ private void processStreamRequest(final StreamRequest req) {
+      long chunksBeingTransferred = streamManager.chunksBeingTransferred();
+      if (chunksBeingTransferred >= maxChunksBeingTransferred) {
+        logger.warn("The number of chunks being transferred {} is above {}, close the connection.",
+            chunksBeingTransferred, maxChunksBeingTransferred);
```
Nit: indentation should be 2 spaces

AngersZhuuuu replied:
> Nit: indentation should be 2 spaces

Done

otterc reviewed (commented):
Other than the nits, looks good to me.

AngersZhuuuu committed:
follow comment

SparkQA commented:
**[Test build #131087 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131087/testReport)** for PR 30139 at commit [`fab5557`](https://github.com/apache/spark/commit/fab5557e1e0a0f4b78868ab98e55b05b997864b2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35690/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35690/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35690/
Test FAILed.

SparkQA commented:
**[Test build #131087 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131087/testReport)** for PR 30139 at commit [`fab5557`](https://github.com/apache/spark/commit/fab5557e1e0a0f4b78868ab98e55b05b997864b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131087/
Test PASSed.

AngersZhuuuu commented:
Any update?

mridulm commented:
+CC @srowen, @HyukjinKwon I merged the pr via ./dev/merge_spark_pr.py - but it failed with following [1]
When I look at apache repo, the commit has made it through : https://git-wip-us.apache.org/repos/asf?p=spark.git;a=commit;h=f8b95dddc1571194fd728d7e0c6de495895da99e

Any idea why there is a disconnect between github and apache repo ? Thanks.


[1]
```
Merge complete (local ref PR_TOOL_MERGE_PR_30139_MASTER). Push to apache? (y/n): y
git push apache PR_TOOL_MERGE_PR_30139_MASTER:master
Username for 'https://git-wip-us.apache.org': mridulm80
Password for 'https://mridulm80@git-wip-us.apache.org': 
Enumerating objects: 26, done.
Counting objects: 100% (26/26), done.
Delta compression using up to 8 threads
Compressing objects: 100% (9/9), done.
Writing objects: 100% (14/14), 4.72 KiB | 4.72 MiB/s, done.
Total 14 (delta 5), reused 0 (delta 0)
remote: To git@github:apache/spark.git
remote:  ! [rejected]        f8b95dddc1571194fd728d7e0c6de495895da99e -> master (fetch first)
remote: error: failed to push some refs to 'git@github:apache/spark.git'
remote: hint: Updates were rejected because the remote contains work that you do
remote: hint: not have locally. This is usually caused by another repository pushing
remote: hint: to the same ref. You may want to first integrate the remote changes
remote: hint: (e.g., 'git pull ...') before pushing again.
remote: hint: See the 'Note about fast-forwards' in 'git push --help' for details.
remote: Syncing refs/heads/master...
remote: Could not sync with GitHub: 
remote: Sending notification emails to: [u'"commits@spark.apache.org" <commits@spark.apache.org>']
remote: Error running hook: /x1/gitbox/hooks/post-receive.d/01-sync-repo.py
To https://git-wip-us.apache.org/repos/asf/spark.git
   5e8549973dc..f8b95dddc15  PR_TOOL_MERGE_PR_30139_MASTER -> master
git rev-parse PR_TOOL_MERGE_PR_30139_MASTER
Restoring head pointer to master
git checkout master
Switched to branch 'master'
git branch
Deleting local branch PR_TOOL_MERGE_PR_30139
git branch -D PR_TOOL_MERGE_PR_30139
Deleting local branch PR_TOOL_MERGE_PR_30139_MASTER
git branch -D PR_TOOL_MERGE_PR_30139_MASTER
Pull request #30139 merged!
Merge hash: f8b95ddd

```

mridulm commented:
Scratch that - was a concurrent merge issue at git, not seen this before :-)

mridulm commented:
Thanks for fixing this @AngersZhuuuu !

Thanks for the reviews @jiangxb1987, @HeartSaVioR, @otterc, @srowen, @cloud-fan

HeartSaVioR commented:
I became a committer after gitbox integration, and according to the guide, I guess we no longer need to set up ASF git as remote repo.

mridulm commented:
@AngersZhuuuu Can you update the jira please ? I am not sure if I have added your id as assignee

mridulm commented:
My repo config's are from ages back @HeartSaVioR :-)
I should perhaps update them ... good point.

AngersZhuuuu commented:
> @AngersZhuuuu Can you update the jira please ? I am not sure if I have added your id as assignee

assignee is me now, any other information need update?
![image](https://user-images.githubusercontent.com/46485123/99477417-6e255500-298d-11eb-9434-10796b29df70.png)

mridulm commented:
Thanks !

HyukjinKwon commented:
Yeah .. I migrated to gitbox completely IIRC. I remember there was a similar syncing issue without the migration. Now committers have to set up according to https://infra.apache.org/apache-github.html if I am not wrong. FYI, I believe we now can use interchangeably gitbox and github. This is my remote FYI:

```bash
git remote -v
```
```
...
apache	https://github.com/apache/spark.git (fetch)
apache	https://github.com/apache/spark.git (push)
apache-github	https://github.com/apache/spark.git (fetch)
apache-github	https://github.com/apache/spark.git (push)
...
origin	https://github.com/HyukjinKwon/spark.git (fetch)
origin	https://github.com/HyukjinKwon/spark.git (push)
...
upstream	https://github.com/apache/spark.git (fetch)
upstream	https://github.com/apache/spark.git (push)
...
```
